include/aero/grid/grid.hpp

```diff
@@ -1,44 +1,54 @@
 #ifndef AERO_GRID_HPP
 #define AERO_GRID_HPP
 
+#include <aero/array/array.hpp>
+
+#include <memory>
+
 namespace aero {
 
-class Grid {
+/// This type represents a computional grid consisting of a set of contiguous
+/// segments separated by interfaces. The grid keeps track of the midpoints
+/// between interfaces.
+struct Grid final {
 public:
-  /// Default constructor.
-  Grid() = default;
-  /// Copy constructor
-  Grid(const Grid& other) = default;
+  // A grid may be neither default constructed nor copied from another grid.
+  Grid() = delete;
+  Grid(const Grid&) = delete;
+  Grid& operator=(const Grid&) = delete;
+
+  /// Creates a grid from an Array containing points delimiting the interfaces
+  /// between segments. The Grid assumes ownership of this array.
+  explicit Grid(Array* interfaces);
+
+  /// Creates a grid from a C pointer. The grid data is copied so it lives in
+  /// both C and C++ runtimes.
+  static Grid from_c_ptr(void *c_ptr);
+
+  /// Creates a grid from a Fortran pointer. The Grid data is copied so it
+  /// lives in both Fortran and C++ runtimes.
+  static Grid from_fortran_ptr(void *f_ptr);
```

Comment:
I guess the biggest advantage to having `Grid` be a non-polymorphic type is that no one will immediately run off and implement a 10-deep `Grid` hierarchy that doesn't need to exist. I do agree with your account of the drawbacks, though.

If we don't want grids to be copied over language boundaries, we could implement them polymorphically and try to reign in the inevitable "oh boy another thing to inherit from!" responses. Maybe we could also think about a way that a Grid could just contain arrays, no matter what they are. I don't really know how important it is that a Grid exclusively owns its arrays on a per-instance basis. Relaxing that constraint would obviate the need for copies.

Reply:
That makes sense, although I guess I'm wondering, given our audience, how likely it will be that people will run amok with inheritance vs. not wanting to touch it with a 10' pole.

I do worry about people not realizing when they are giving up control of an object though, and it makes me a little worried about something like having a `Grid` take ownership (or expect the persistence) of an `Array`.